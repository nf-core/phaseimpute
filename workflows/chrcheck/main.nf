include { VCFCHREXTRACT as VCFCHRBF } from '../../modules/local/vcfchrextract'
include { BAMCHREXTRACT as BAMCHRBF } from '../../modules/local/bamchrextract'
include { BAM_CHR_RENAME_SAMTOOLS   } from '../../subworkflows/local/bam_chr_rename_samtools'
include { VCF_CHR_RENAME_BCFTOOLS   } from '../../subworkflows/local/vcf_chr_rename_bcftools'
include { VCFCHREXTRACT as VCFCHRAF } from '../../modules/local/vcfchrextract'
include { BAMCHREXTRACT as BAMCHRAF } from '../../modules/local/bamchrextract'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

workflow CHR_CHECK {
    take:
        ch_input // [[id], file, index]
        ch_fasta // [[id], fasta, fai]

    main:
        ch_versions = Channel.empty()

        // Split the input between VCF and BAM files
        ch_input = ch_input.branch{
            bam: it[1] =~ 'bam|cram|sam'
            vcf: it[1] =~ 'vcf|bcf'
        }
        ch_input.vcf.view()
        ch_input.bam.view()

        // Extract the contig names from the VCF files
        VCFCHRBF(ch_input.vcf)
        ch_versions = ch_versions.mix(VCFCHRBF.out.versions)
        chr_vcf_disjoint = check_chr(VCFCHRBF.out.chr, ch_input.vcf, ch_fasta)

        // Extract the contig names from the BAM files
        BAMCHRBF(ch_input.bam)
        ch_versions = ch_versions.mix(BAMCHRBF.out.versions)
        chr_bam_disjoint = check_chr(BAMCHRBF.out.chr, ch_input.bam, ch_fasta)

        if (params.rename_chr == true) {
            // Rename the contigs in the BAM files
            BAM_CHR_RENAME_SAMTOOLS(
                chr_bam_disjoint.to_rename.map{meta, bam, csi, dis, prefix -> [meta, bam, csi, prefix]}
            )
            ch_versions = ch_versions.mix(BAM_CHR_RENAME_SAMTOOLS.out.versions)

            // Rename the contigs in the VCF files
            VCF_CHR_RENAME_BCFTOOLS(
                chr_vcf_disjoint.to_rename.map{meta, vcf, csi, dis, prefix -> [meta, vcf, csi]}
            )

            // Check if modification has solved the problem
            BAMCHRAF(BAM_CHR_RENAME_SAMTOOLS.out.bam_renamed.map{ meta, bam, csi -> [meta, bam] })
            chr_bam_disjoint_after = check_chr(BAMCHRAF.out.chr, BAM_CHR_RENAME_SAMTOOLS.out.bam_renamed, ch_fasta)

            VCFCHRAF(VCF_CHR_RENAME_BCFTOOLS.out.vcf_renamed.map{ meta, vcf, csi -> [meta, vcf] })
            chr_vcf_disjoint_after = check_chr(VCFCHRAF.out.chr, VCF_CHR_RENAME_BCFTOOLS.out.vcf_renamed, ch_fasta)

            chr_bam_disjoint_after.to_rename.map{
                error "Even after renaming errors are still present. Please check the contigs name: ${it[3]} in BAM: ${it[1]} and fasta file."
            }
            chr_vcf_disjoint_after.to_rename.map{
                error "Even after renaming errors are still present. Please check the contigs name: ${it[3]} in VCF: ${it[1]} and fasta file."
            }
            // If no errors are present, we can use the renamed files
            ch_bam_renamed = BAM_CHR_RENAME_SAMTOOLS.out.bam_renamed
            ch_vcf_renamed = VCF_CHR_RENAME_BCFTOOLS.out.vcf_renamed
        } else {
            chr_vcf_disjoint.to_rename.map {
                error "Contig names: ${it[3]} in the VCF: ${it[1]} are not present in the reference genome. Please set `rename_chr` to `true` to rename the contigs."
            }
            chr_bam_disjoint.to_rename.map {
                error "Contig names: ${it[3]} in the BAM: ${it[1]} are not present in the reference genome. Please set `rename_chr` to `true` to rename the contigs."
            }
            ch_vcf_renamed = Channel.empty()
            ch_bam_renamed = Channel.empty()
        }

        ch_output = chr_bam_disjoint.no_rename
            .mix(chr_vcf_disjoint.no_rename)
            .mix(ch_bam_renamed)
            .mix(ch_vcf_renamed)
    emit:
        output   = ch_output             // [ [id], file, index ]
        versions = ch_versions           // channel: [ versions.yml ]
}


def check_chr(ch_chr, ch_input, ch_fasta){
    chr_checked = ch_chr
        .combine(ch_input, by:0)
        .combine(ch_fasta)
        .map{metaI, chr, file, index, metaG, fasta, fai ->
            [
                metaI, file, index,
                chr.readLines()*.split(' ').collect{it[0]},
                fai.readLines()*.split('\t').collect{it[0]}
            ]
        }
        .branch{ meta, file, index, chr, fai ->
            no_rename: (chr - fai).size() == 0
                return [meta, file, index]
            to_rename: true
                return [meta, file, index, chr-fai, chr-fai =~ "chr" ? "nochr" : "chr"]
        }
    return chr_checked
}
