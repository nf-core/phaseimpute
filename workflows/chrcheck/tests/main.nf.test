nextflow_workflow {

    name "Test workflow CHRCHECK"
    script "../main.nf"

    workflow "CHRCHECK"

    tag "workflows"
    tag "workflows_local"
    tag "workflows/chrcheck"
    tag "chrcheck"

    tag "bcftools"
    tag "bcftools/annotate"
    tag "bcftools/index"
    tag "gawk"
    tag "samtools"
    tag "samtools/reheader"
    tag "samtools/index"

    test("Rename: VCF BAM chr + fasta no chr") {
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "VCF_chr22"],
                        file(params.test_data['homo_sapiens']['genome']['dbsnp_146_hg38_vcf_gz'],checkIfExist:true),
                        file(params.test_data['homo_sapiens']['genome']['dbsnp_146_hg38_vcf_gz_tbi'],checkIfExist:true),
                        ["22"]
                    ],
                    [
                        [id: "BAM_chr22"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true),
                        ["22"]
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.output.collect{
                    if (it[1].endsWith("vcf.gz")) {
                        path(it[1]).linesGzip.findAll { it.startsWith ("##contig=<ID") }
                    } else {
                        bam(it[1]).getHeader().findAll { it.startsWith ("@SQ") }
                    }
                    }).match("headernochr")
                }
            )
        }
    }

    test("Rename: VCF no chr + fasta chr") {
        config "./nextflow_rename.config"
        setup {
            run("BAM_CHR_RENAME_SAMTOOLS", alias: "PREPROCESS") {
                script "../../../subworkflows/local/bam_chr_rename_samtools/main.nf"
                process {
                    """
                    input[0] = Channel.fromList([
                        [
                            [id: "BAM_22"],
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                            file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true),
                            "nochr"
                        ]
                    ])
                    """
                }
            }
        }
        when {
            workflow {
                """
                input[0] = PREPROCESS.out.bam_renamed
                    .mix(
                        Channel.fromList([
                            [
                                [id: "VCF_AllNoChr"],
                                file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/NA24385_sv.vcf.gz",checkIfExist:true),
                                file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/NA24385_sv.vcf.gz.tbi",checkIfExist:true),
                            ],
                        ])
                    )
                    .combine(Channel.of(["chr22"]).collect().toList())
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() },
                { assert snapshot(workflow.out.output.collect{
                    if (it[1].endsWith("vcf.gz")) {
                        path(it[1]).linesGzip.findAll { it.startsWith ("##contig=<ID") }
                    } else {
                        bam(it[1]).getHeader().findAll { it.startsWith ("@SQ") }
                    }
                    }).match("headerwithchr")
                }
            )
        }
    }

    test("Error : missing renaming VCF") {
        config "./nextflow.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "VCF_chr22"],
                        file(params.test_data['homo_sapiens']['genome']['dbsnp_146_hg38_vcf_gz'],checkIfExist:true),
                        file(params.test_data['homo_sapiens']['genome']['dbsnp_146_hg38_vcf_gz_tbi'],checkIfExist:true),
                        ["22"]
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
                { assert workflow.errorReport
                    .contains("Contig names: [chr22] in VCF: /nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz are not present in reference genome with same writing. Please set `rename_chr` to `true` to rename the contigs.")
                }
            )
        }
    }

    test("Error : missing renaming BAM") {
        config "./nextflow.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "BAM_chr22"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true),
                        ["22"]
                    ]
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
                { assert workflow.errorReport
                    .contains("Contig names: [chr22] in BAM: /nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam are not present in reference genome with same writing. Please set `rename_chr` to `true` to rename the contigs.")
                }
            )
        }
    }

    test("Error : still difference after renaming VCF"){
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                lst_chr = [
                    "chr1", "chr2", "chr3", "chr4", "chr5", "chr6", "chr7", "chr8", "chr9", "chr10", "chr34",
                    "GL000207.1", "GL000226.1", "GL000229.1", "GL000231.1", "GL000210.1", "GL000239.1", "GL000235.1",
                    "GL000201.1", "GL000247.1", "GL000245.1", "GL000197.1", "GL000203.1", "GL000246.1", "GL000249.1",
                    "GL000196.1", "GL000248.1", "GL000244.1", "GL000238.1", "GL000202.1", "GL000234.1", "GL000232.1",
                    "GL000206.1", "GL000240.1", "GL000236.1", "GL000241.1", "GL000243.1", "GL000242.1", "GL000230.1",
                    "GL000237.1", "GL000233.1", "GL000204.1", "GL000198.1", "GL000208.1", "GL000191.1", "GL000227.1",
                    "GL000228.1", "GL000214.1", "GL000221.1", "GL000209.1", "GL000218.1", "GL000220.1", "GL000213.1",
                    "GL000211.1", "GL000199.1", "GL000217.1", "GL000216.1", "GL000215.1", "GL000205.1", "GL000219.1",
                    "GL000224.1", "GL000223.1", "GL000195.1", "GL000212.1", "GL000222.1", "GL000200.1", "GL000193.1",
                    "GL000194.1", "GL000225.1", "GL000192.1", "NC_007605", "hs37d5"
                ]
                input[0] = Channel.fromList([
                    [
                        [id: "VCF_AllNoChr"],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/NA24385_sv.vcf.gz",checkIfExist:true),
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/NA24385_sv.vcf.gz.tbi",checkIfExist:true),
                        lst_chr
                    ],
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
                { assert workflow.errorReport.contains("Contig names: [chr34] absent from file: /nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/NA24385_sv.vcf.gz and cannot be solved by adding or removing the `chr` prefix.")}
            )
        }
    }

    test("Error : still difference after renaming BAM"){
        config "./nextflow_rename.config"
        when {
            workflow {
                """
                input[0] = Channel.fromList([
                    [
                        [id: "BAM_chr22"],
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam'], checkIfExist:true),
                        file(params.test_data['homo_sapiens']['illumina']['test_paired_end_sorted_bam_bai'], checkIfExist:true),
                        ["chr1"]
                    ],
                ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.failed },
                { assert workflow.errorReport.contains("Contig names: [chr1] absent from file: /nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/bam/test.paired_end.sorted.bam and cannot be solved by adding or removing the `chr` prefix.")}
            )
        }
    }
}
