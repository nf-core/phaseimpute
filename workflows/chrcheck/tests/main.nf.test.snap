{
    "Rename: VCF chr + fasta chr": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "chr21"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi"
                    ],
                    [
                        {
                            "id": "chr22"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi"
                    ]
                ],
                "1": [
                    "versions.yml:md5,1d193f9fc9c33a25b3515ee3aa464399",
                    "versions.yml:md5,1d193f9fc9c33a25b3515ee3aa464399"
                ],
                "output": [
                    [
                        {
                            "id": "chr21"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi"
                    ],
                    [
                        {
                            "id": "chr22"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,1d193f9fc9c33a25b3515ee3aa464399",
                    "versions.yml:md5,1d193f9fc9c33a25b3515ee3aa464399"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-06-19T16:22:17.949552063"
    },
    "Rename: panel chr + fasta chr": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "chr21"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi"
                    ],
                    [
                        {
                            "id": "chr22"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi"
                    ]
                ],
                "1": [
                    "versions.yml:md5,395e1cde3f38a30f5d80769972ba23d8",
                    "versions.yml:md5,ad4c5338cd27e20789c70e28b8c74a42"
                ],
                "vcf": [
                    [
                        {
                            "id": "chr21"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/illumina/vcf/chr21/simulated_sv.vcf.gz.tbi"
                    ],
                    [
                        {
                            "id": "chr22"
                        },
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz",
                        "/nf-core/test-datasets/modules/data/genomics/homo_sapiens/genome/vcf/dbsnp_146.hg38.vcf.gz.tbi"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,395e1cde3f38a30f5d80769972ba23d8",
                    "versions.yml:md5,ad4c5338cd27e20789c70e28b8c74a42"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-27T17:21:13.588561053"
    },
    "Rename: panel no chr + fasta chr": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "22"
                        },
                        "22_chrrename.vcf.gz:md5,070a96d1053a64f2de2132ee8800847c",
                        "22_chrrename.vcf.gz.csi:md5,e190b690b4b0a4d088231862e5408582"
                    ]
                ],
                "1": [
                    "versions.yml:md5,395e1cde3f38a30f5d80769972ba23d8",
                    "versions.yml:md5,ad4c5338cd27e20789c70e28b8c74a42",
                    "versions.yml:md5,e576f40503c3506c782228485d06fbf1"
                ],
                "vcf": [
                    [
                        {
                            "id": "22"
                        },
                        "22_chrrename.vcf.gz:md5,070a96d1053a64f2de2132ee8800847c",
                        "22_chrrename.vcf.gz.csi:md5,e190b690b4b0a4d088231862e5408582"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,395e1cde3f38a30f5d80769972ba23d8",
                    "versions.yml:md5,ad4c5338cd27e20789c70e28b8c74a42",
                    "versions.yml:md5,e576f40503c3506c782228485d06fbf1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-27T17:21:39.92481538"
    },
    "Rename: panel chr + fasta no chr": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "chr21"
                        },
                        "chr21_chrrename.vcf.gz:md5,22785a5d7ec1132f766efae5f8e00adf",
                        "chr21_chrrename.vcf.gz.csi:md5,b5b5fd753ee54ebd3c8e4b1fe2261cdb"
                    ],
                    [
                        {
                            "id": "chr22"
                        },
                        "chr22_chrrename.vcf.gz:md5,23de9b4db1406415806e627969cec749",
                        "chr22_chrrename.vcf.gz.csi:md5,ba370ca13289fee4be59253a1f4609e2"
                    ]
                ],
                "1": [
                    "versions.yml:md5,395e1cde3f38a30f5d80769972ba23d8",
                    "versions.yml:md5,ad4c5338cd27e20789c70e28b8c74a42",
                    "versions.yml:md5,e576f40503c3506c782228485d06fbf1"
                ],
                "vcf": [
                    [
                        {
                            "id": "chr21"
                        },
                        "chr21_chrrename.vcf.gz:md5,22785a5d7ec1132f766efae5f8e00adf",
                        "chr21_chrrename.vcf.gz.csi:md5,b5b5fd753ee54ebd3c8e4b1fe2261cdb"
                    ],
                    [
                        {
                            "id": "chr22"
                        },
                        "chr22_chrrename.vcf.gz:md5,23de9b4db1406415806e627969cec749",
                        "chr22_chrrename.vcf.gz.csi:md5,ba370ca13289fee4be59253a1f4609e2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,395e1cde3f38a30f5d80769972ba23d8",
                    "versions.yml:md5,ad4c5338cd27e20789c70e28b8c74a42",
                    "versions.yml:md5,e576f40503c3506c782228485d06fbf1"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-03-27T17:21:28.214969089"
    }
}